js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx

```diff
@@ -1,118 +1,181 @@
-import {ApolloQueryResult, DocumentNode, OperationVariables, useApolloClient} from '@apollo/client';
+import {ApolloClient, DocumentNode, OperationVariables, useApolloClient} from '@apollo/client';
 import {cache} from 'idb-lru-cache';
-import React from 'react';
+import memoize from 'lodash/memoize';
+import React, {createContext, useCallback, useContext} from 'react';
 
 type CacheData<TQuery> = {
   data: TQuery;
   version: number;
 };
 
-let fetchState: Record<
-  string,
-  {
-    onFetched: ((value: any) => void)[];
+export const KEY_PREFIX = 'indexdbQueryCache:';
+
+export class CacheManager<TQuery> {
+  private cache: ReturnType<typeof cache<string, CacheData<TQuery>>>;
+  private key: string;
+
+  constructor(key: string) {
+    this.key = `${KEY_PREFIX}${key}`;
+    this.cache = cache<string, CacheData<TQuery>>({dbName: this.key, maxCount: 1});
   }
-> = {};
 
-/**
- * Returns data from the indexedDB cache initially while loading is true.
- * Fetches data from the network/cache initially and does not receive any updates afterwards
- * Uses fetch-policy: no-cache to avoid slow apollo cache normalization
- */
+  async get(version: number): Promise<TQuery | null> {
+    if (await this.cache.has('cache')) {
+      const {value} = await this.cache.get('cache');
+      if (value && version === value.version) {
+        return value.data;
+      }
+    }
+    return null;
+  }
+
+  set(data: TQuery, version: number): Promise<void> {
+    return this.cache.set('cache', {data, version}, {expiry: new Date('3030-01-01')});
+  }
+}
+
+interface QueryHookParams<TVariables extends OperationVariables, TQuery> {
+  key: string;
+  query: DocumentNode;
+  version: number;
+  variables?: TVariables;
+  onCompleted?: (data: TQuery) => void;
+}
+
 export function useIndexedDBCachedQuery<TQuery, TVariables extends OperationVariables>({
   key,
   query,
   version,
   variables,
-}: {
-  key: string;
-  query: DocumentNode;
-  version: number;
-  variables?: TVariables;
-}) {
+}: QueryHookParams<TVariables, TQuery>) {
```

Comment:
Another option here might be to use the Cache APIs to store this data with the rest of the cached of images/js etc https://developer.mozilla.org/en-US/docs/Web/API/Cache

Reply:
Yeah I believe @alangenfeld is going to research how we make these queries cached. For using the Cache API I didn't realize it was available outside of service workers!
> You don't have to use it in conjunction with service workers, even though it is defined in the service worker spec.

We could totally use a service worker to keep the workspace up to date

Reply:
>I'm a little bit worried about this API being so nice and easy that we use useIndexedDBCachedQuery a lot, and create tiny never-deleted indexeddb databases for each query. For less-critical cached data, using a single indexeddb with keys + max key count etc (eg: actually using idb-lru-cache as an LRU cache) might be nicer on the user's computer.

Yeah this is true, It might be nice to generalize this to cache more queries, for now though we're only using it to cache definition level queries that currently pull the workspace snapshot on the backend. For other queries I don't think it's too important too cache them in indexeddb but if we wanted to for those I believe we should tie it to the apollo cache (dump the apollo cache in indexeddb and then rehydrate it from indexeddb when the page loads)